modules/nf-core/limma/differential/templates/limma_de_rnaseq.R

```diff
@@ -0,0 +1,411 @@
+        contrast <- paste(paste(contrast_variable, c(opt\$reference_level, opt\$target_level), sep='.'), collapse='-')
+        contrast_matrix <- makeContrasts(contrasts=contrast, levels=design)
+        fit <- contrasts.fit(fit, contrast_matrix)
+    }
+
+    ebayes_args = list(fit = fit)
+    fit <- do.call(eBayes, ebayes_args)
+
+} else if (opt\$analysis_type == "mixedmodel") {
+
+    ## CHECK CONTRAST SPECIFICATION
+    contrast_variable <- make.names(opt\$contrast_variable)
+    reference_level <- make.names(opt\$reference_level)
+    target_level <- make.names(opt\$target_level)
+    sample.sheet[[contrast_variable]] = paste(sample.sheet[[strsplit2(contrast_variable, split = "\\\\.")[1]]],
+         sample.sheet[[strsplit2(contrast_variable, split = "\\\\.")[2]]], sep = ".")
+
+    validate_sample_sheet(sample.sheet, contrast_variable, reference_level, target_level, opt\$blocking_variables)
+
+    blocking.vars <- sample.sheet[[opt\$blocking_variables]]
+
+    # Mixed model analysis setup
+    model <- paste('~ 0 +', contrast_variable)
+    design <- model.matrix(as.formula(model), data=sample.sheet)
+    colnames(design) <- sub(contrast_variable, paste0(contrast_variable, '.'), colnames(design))
+
+    voom_args = list(counts = DGE, design = design, plot = FALSE)
+    voom_1 <- do.call(voom, voom_args)
+        # Write the normalized counts matrix to a TSV file
+    normalized_counts <- voom_1\$E
+    normalized_counts_with_genes <- data.frame(Gene = rownames(normalized_counts), normalized_counts, row.names = NULL)
+    colnames(normalized_counts_with_genes)[1] <- opt\$probe_id_col
+    write.table(normalized_counts_with_genes,
+        file = paste(opt\$output_prefix, "normalised_counts.tsv", sep = '.'),
+        sep = "\t",
+        quote = FALSE,
+        row.names = FALSE)
+
+    corfit_args <- list(object = voom_1, design = design, block = blocking.vars)
+    corfit = do.call(duplicateCorrelation, corfit_args)
+
+    voom_args <- list(counts = DGE, design = design, plot = FALSE, correlation = corfit\$consensus.correlation)
+    voom_1 <- do.call(voom, voom_args)
+
+    corfit_args <- list(object = voom_1, design = design, block = blocking.vars)
+    corfit = do.call(duplicateCorrelation, corfit_args)
+
+    lmfit_args = list(object = voom_1, design = design, block = blocking.vars, correlation = corfit\$consensus.correlation)
+    fit <- do.call(lmFit, lmfit_args)
+
+    if (opt\$do_contrast) {
+        contrast <- paste(paste(contrast_variable, c(reference_level, target_level), sep='.'), collapse='-')
+        contrast_matrix <- makeContrasts(contrasts=contrast, levels=design)
+        fit <- contrasts.fit(fit, contrast_matrix)
+    }
+
+    ebayes_args = list(fit = fit)
+    fit <- do.call(eBayes, ebayes_args)
+
+} else {
+    stop(
+        paste("Bad '--analysis_type' parameter. It can be either 'paired' or 'mixedmodel'. ")
+    )
+}
+
+## Generate outputs
+# Differential expression table
+
+if (opt\$IHW_correction) {
```

Comment:
IMO if we're writing 'limma' modules they should be thin-as-possible wrappers around Limma functions. We shouldn't we embedding lots of additional logic, aside from the boilerplate required to pass inputs to those functions correctly and write the outputs.

Reply:
Extensive customisations should probably be in local modules in the workflow, and we'd want to discuss to determine if this was part of standard best practice.